sources/core/Stride.Core.CompilerServices/Analyzers/STRDIAG010InvalidConstructor.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Stride.Core.CompilerServices.Common;
+
+namespace Stride.Core.CompilerServices.Analyzers;
+
+[DiagnosticAnalyzer(LanguageNames.CSharp)]
+public class STRDIAG010InvalidConstructor : DiagnosticAnalyzer
+{
+    public const string DiagnosticId = "STRDIAG010";
+    private const string Title = "Invalid Constructor";
+    private const string MessageFormat = "The Type '{0}' doesn't have an empty constructor, which is needed for Serialization";
+    private const string Category = DiagnosticCategory.Serialization;
+
+    private static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+        DiagnosticId,
+        Title,
+        MessageFormat,
+        Category,
+        DiagnosticSeverity.Warning,
+        isEnabledByDefault: true,
+        helpLinkUri: string.Format(DiagnosticCategory.LinkFormat, DiagnosticId));
+
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get { return ImmutableArray.Create(Rule); } }
+
+    public override void Initialize(AnalysisContext context)
+    {
+        context.EnableConcurrentExecution();
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+        context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+    }
+
+    private static void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+    {
+        var dataContractAttribute = WellKnownReferences.DataContractAttribute(context.Compilation);
+
+
+        if (dataContractAttribute is null)
+            return;
+
+        context.RegisterSymbolAction(symbolContext => AnalyzeSymbol(symbolContext, dataContractAttribute), SymbolKind.NamedType);
+
+    }
+
+    private static void AnalyzeSymbol(SymbolAnalysisContext context, INamedTypeSymbol dataContractAttribute)
+    {
+        var symbol = (INamedTypeSymbol)context.Symbol;
+        if (symbol.IsAbstract)
+            return;
+
+        if (symbol.HasAttribute(dataContractAttribute))
+        {
+            TryReportDiagnostics(symbol, context);
+        }
+
+        var type = symbol.BaseType;
+        bool isInherited = false;
+        while (type != null)
+        {
+            // Check if the type has the specified DataContractAttribute through inheritance
+            if (type.TryGetAttribute(dataContractAttribute,out var datacontractData) && datacontractData.AttributeConstructor is not null)
+            {
+                if (datacontractData is { NamedArguments: [.., { Key: "Inherited" } ], NamedArguments: [..,{ Value: TypedConstant  inherited }] })
```

Comment:
Is this pattern matching syntax based on index of the argument (second in this case)? I think it would be better to use something like `TryGetValue` pattern if supported by `NamedArguments` if the index can change

Reply:
https://github.com/stride3d/stride/pull/2104/commits/84c1376c00bb0864f49e8293caa59bb1583531de
added a test to confirm it works

the pattern matching looks "loop until Inherited appears, then put it inot a bool" and fills it up so thats fine

Reply:
Would it work to inline Value decomposition into the object earlier?
```
{ NamedArguments: [.., { Key: "Inherited", Value: value }] }
```

Reply:
yes that worked [4e81a8f](https://github.com/stride3d/stride/pull/2104/commits/4e81a8f24f036a3f066858c82b4607276b1afd5d)